pkg/splunk/enterprise/workerpool_test.go

```diff
@@ -0,0 +1,322 @@
+	localPath := filepath.Join(splcommon.AppDownloadVolume, "downloadedApps", "default" /*namespace*/, "Standalone", crName, scope, appSrcName) + "/"
+	err = createAppDownloadDir(localPath)
+	if err != nil {
+		t.Errorf("Unable to create the download directory")
+	}
+	defer os.RemoveAll(splcommon.AppDownloadVolume)
+
+	for idx := 0; idx < len(appNames); idx++ {
+		// create the dummy app packages locally
+		appFileName := appNames[idx] + "_" + objectHashes[idx]
+		appLoc := filepath.Join(localPath, appFileName)
+		err := createOrTruncateAppFileLocally(appLoc, sizes[idx])
+		if err != nil {
+			t.Errorf("Unable to create the app files locally")
+		}
+		defer os.Remove(appLoc)
+	}
+
+	// Update the GetS3Client with our mock call which initializes mock AWS client
+	getClientWrapper := splclient.S3Clients[vol.Provider]
+	getClientWrapper.SetS3ClientFuncPtr(vol.Provider, splclient.NewMockAWSS3Client)
+
+	initFunc := getClientWrapper.GetS3ClientInitFuncPtr()
+
+	s3ClientMgr := S3ClientManager{
+		client:          client,
+		cr:              &standalone,
+		appFrameworkRef: &appFrameworkConfig,
+		vol:             &vol,
+		location:        appSrc.Location,
+		initFn:          initFunc,
+		getS3Client:     GetRemoteStorageClient,
+	}
+
+	workPool := NewAppInstallWorkerPool(splcommon.DefaultMaxConcurrentAppDownloads, appDeployInfoList,
+		s3ClientMgr, s3Response, 200 /*dummy available disk space*/, localPath, scope)
+
+	workPool.Start()
+
+	// check if all apps are downloaded
+	if ok, err := areAppsDownloadedSuccessfully(appDeployInfoList); !ok {
+		t.Errorf("All apps should have been downloaded successfully, error=%v", err)
+	}
+
+	// check again, just for code coverage
+	workPool.Start()
+	// check if all apps are downloaded
+	if ok, err := areAppsDownloadedSuccessfully(appDeployInfoList); !ok {
+		t.Errorf("All apps should have been downloaded successfully, error=%v", err)
+	}
+
+	// Reset the download state for one of the apps, eventually this should be marked DownloadComplete.
+	// We will not re-download it again as it was already downloaded before.
+	workPool = NewAppInstallWorkerPool(splcommon.DefaultMaxConcurrentAppDownloads, appDeployInfoList,
+		s3ClientMgr, s3Response, 200 /*dummy available disk space*/, localPath, scope)
+
+	setAppDownloadState(workPool.JobList[0], enterpriseApi.DownloadNotStarted)
+
+	workPool.Start()
+	// check if all apps are downloaded
+	if ok, err := areAppsDownloadedSuccessfully(appDeployInfoList); !ok {
+		t.Errorf("All apps should have been downloaded successfully, error=%v", err)
+	}
+
+	// remove the apps locally to test download again
+	for idx := 0; idx < len(appNames); idx++ {
+		appFileName := appNames[idx] + "_" + objectHashes[idx]
+		appLoc := filepath.Join(localPath, appFileName)
+		os.Remove(appLoc)
+	}
+
+	// Reduce the availableDiskSpace to test scenario where we dont have enough disk space.
+	workPool = NewAppInstallWorkerPool(splcommon.DefaultMaxConcurrentAppDownloads, appDeployInfoList,
```

Comment:
Is this testing the case when we have enough space to download and push only one app at a time so they have to be queued up serially?

Reply:
This is testing the case where we are able to push few apps and then for the rest of them we have to wait till we have enough space. And then once we have the available space, we will pick them up from the channel